src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeWriter.java

```diff
@@ -0,0 +1,766 @@
+import org.apache.lucene.store.IndexOutput;
+import org.apache.lucene.util.ArrayUtil;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.FutureArrays;
+import org.apache.lucene.util.IntroSorter;
+import org.apache.lucene.util.Sorter;
+import org.apache.lucene.util.bkd.MutablePointsReaderUtils;
+
+/**
+ * This is a specialisation of the lucene BKDWriter that only writes a single dimension
+ * balanced tree.
+ * <p>
+ * Recursively builds a block balanced tree to assign all incoming points to smaller
+ * and smaller rectangles (cells) until the number of points in a given
+ * rectangle is &lt;= <code>maxPointsInLeafNode</code>.  The tree is
+ * fully balanced, which means the leaf nodes will have between 50% and 100% of
+ * the requested <code>maxPointsInLeafNode</code>.  Values that fall exactly
+ * on a cell boundary may be in either cell.
+ *
+ * <p>
+ * <b>NOTE</b>: This can write at most Integer.MAX_VALUE * <code>maxPointsInLeafNode</code> total points.
+ */
+public class BlockBalancedTreeWriter
+{
+    // Enable to check that values are added to the tree in correct order and within bounds
+    private static final boolean DEBUG = false;
+
+    // Default maximum number of point in each leaf block
+    public static final int DEFAULT_MAX_POINTS_IN_LEAF_NODE = 1024;
+
+    private final int bytesPerValue;
+    private final BytesRef scratchBytesRef1 = new BytesRef();
+    private final int maxPointsInLeafNode;
+    private final byte[] minPackedValue;
+    private final byte[] maxPackedValue;
+    private long pointCount;
+    private final long maxDoc;
+
+    public BlockBalancedTreeWriter(long maxDoc, int bytesPerValue, int maxPointsInLeafNode)
+    {
+        if (maxPointsInLeafNode <= 0)
+            throw new IllegalArgumentException("maxPointsInLeafNode must be > 0; got " + maxPointsInLeafNode);
+        if (maxPointsInLeafNode > ArrayUtil.MAX_ARRAY_LENGTH)
+            throw new IllegalArgumentException("maxPointsInLeafNode must be <= ArrayUtil.MAX_ARRAY_LENGTH (= " +
+                                               ArrayUtil.MAX_ARRAY_LENGTH + "); got " + maxPointsInLeafNode);
+
+        this.maxPointsInLeafNode = maxPointsInLeafNode;
+        this.bytesPerValue = bytesPerValue;
+        this.maxDoc = maxDoc;
+
+        minPackedValue = new byte[bytesPerValue];
+        maxPackedValue = new byte[bytesPerValue];
+    }
+
+    public long getPointCount()
+    {
+        return pointCount;
+    }
+
+    public int getBytesPerValue()
+    {
+        return bytesPerValue;
+    }
+
+    public int getMaxPointsInLeafNode()
+    {
+        return maxPointsInLeafNode;
+    }
+
+    /**
+     * Write the point values from a {@link IntersectingPointValues}. The points can be reordered before writing
+     * to disk and does not use transient disk for reordering.
+     */
+    public long writeField(IndexOutput out, IntersectingPointValues reader,
+                           final Callback callback) throws IOException
```

Comment:
Possible alternative signature:

```
public long writeTree(IndexOutput out, IntersectingPointValues reader, Callback callback) throws IOException
```

Or it could just be `write()`. I guess we know what we're writing without further qualification, right? :)